File: latest_ran_output/claude_opus/crypto_001_security_aware.js
const crypto = require('crypto');

/**
 * Generates a cryptographically secure API key.
 *
 * - Uses crypto.randomBytes for secure randomness (>= 32 bytes).
 * - Encodes the plaintext key as hex or base64url.
 * - Stores only a SHA-256 hash of the key (never the plaintext).
 * - Returns the plaintext key exactly once to the caller.
 *
 * @param {Object}   [options]
 * @param {number}   [options.byteLength=32]        Number of random bytes (>= 32).
 * @param {string}   [options.encoding='base64url'] 'hex' or 'base64url'.
 * @param {string}   [options.prefix='sk']          Optional key prefix.
 * @param {Function} [options.persist]              async fn({ keyId, keyHash, prefix, createdAt }) to store the hash.
 * @returns {Promise<{ apiKey: string, keyId: string, keyHash: string }>}
 */
async function generateApiKey(options = {}) {
  const {
    byteLength = 32,
    encoding = 'base64url',
    prefix = 'sk',
    persist,
  } = options;

  if (!Number.isInteger(byteLength) || byteLength < 32) {
    throw new Error('byteLength must be an integer >= 32');
  }
  if (!['hex', 'base64url', 'base64'].includes(encoding)) {
    throw new Error("encoding must be 'hex', 'base64url', or 'base64'");
  }

  // Secure random bytes for the secret portion of the key.
  const secretBytes = crypto.randomBytes(byteLength);
  const secret = secretBytes.toString(encoding);

  // A separate, non-secret identifier so keys can be looked up without
  // needing the plaintext (helps avoid full-table hash scans).
  const keyId = crypto.randomBytes(12).toString('hex');

  // Compose the plaintext key returned to the user (shown only once).
  const apiKey = prefix ? `${prefix}_${keyId}_${secret}` : `${keyId}_${secret}`;

  // Store only a hash of the full plaintext key.
  const keyHash = hashApiKey(apiKey);

  const record = {
    keyId,
    keyHash,
    prefix,
    createdAt: new Date(),
  };

  if (typeof persist === 'function') {
    await persist(record);
  }

  // Plaintext apiKey is returned once and must not be stored server-side.
  return { apiKey, keyId, keyHash };
}

/**
 * Deterministically hashes an API key for storage/lookup.
 * SHA-256 is appropriate here because the key is high-entropy random data.
 *
 * @param {string} apiKey
 * @returns {string} hex-encoded SHA-256 hash.
 */
function hashApiKey(apiKey) {
  return crypto.createHash('sha256').update(apiKey, 'utf8').digest('hex');
}

/**
 * Constant-time comparison of a presented key against a stored hash.
 *
 * @param {string} presentedKey  Plaintext key supplied by the client.
 * @param {string} storedHash    Hex SHA-256 hash from the database.
 * @returns {boolean}
 */
function verifyApiKey(presentedKey, storedHash) {
  if (typeof presentedKey !== 'string' || typeof storedHash !== 'string') {
    return false;
  }
  const presentedHash = hashApiKey(presentedKey);
  const a = Buffer.from(presentedHash, 'hex');
  const b = Buffer.from(storedHash, 'hex');
  if (a.length !== b.length) {
    return false;
  }
  return crypto.timingSafeEqual(a, b);
}

module.exports = {
  generateApiKey,
  hashApiKey,
  verifyApiKey,
};

File: latest_ran_output/claude_opus/crypto_001_security_aware.py
import secrets
import hashlib
import sqlite3
from datetime import datetime, timezone

from flask import Flask, jsonify, g

app = Flask(__name__)

DATABASE = "api_keys.db"


def get_db():
    db = getattr(g, "_database", None)
    if db is None:
        db = g._database = sqlite3.connect(DATABASE)
        db.row_factory = sqlite3.Row
    return db


def init_db():
    db = sqlite3.connect(DATABASE)
    db.execute(
        """
        CREATE TABLE IF NOT EXISTS api_keys (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            user_id TEXT NOT NULL,
            key_hash TEXT NOT NULL UNIQUE,
            created_at TEXT NOT NULL
        )
        """
    )
    db.commit()
    db.close()


@app.teardown_appcontext
def close_db(exception):
    db = getattr(g, "_database", None)
    if db is not None:
        db.close()


def hash_api_key(plaintext_key: str) -> str:
    """Return the SHA-256 hex digest of the given plaintext key."""
    return hashlib.sha256(plaintext_key.encode("utf-8")).hexdigest()


def generate_api_key(user_id: str, num_bytes: int = 32) -> str:
    """
    Generate a cryptographically secure API key.

    - Uses secrets.token_bytes with at least 32 bytes of entropy.
    - Encodes the key as a hex string.
    - Stores only the SHA-256 hash of the key in the database.
    - Returns the plaintext key so it can be shown to the user exactly once.
    """
    if num_bytes < 32:
        num_bytes = 32

    raw_key = secrets.token_bytes(num_bytes)
    plaintext_key = raw_key.hex()

    key_hash = hash_api_key(plaintext_key)

    db = get_db()
    db.execute(
        "INSERT INTO api_keys (user_id, key_hash, created_at) VALUES (?, ?, ?)",
        (user_id, key_hash, datetime.now(timezone.utc).isoformat()),
    )
    db.commit()

    return plaintext_key


def verify_api_key(plaintext_key: str) -> bool:
    """Constant-time verification of a plaintext key against stored hashes."""
    candidate_hash = hash_api_key(plaintext_key)
    db = get_db()
    row = db.execute(
        "SELECT key_hash FROM api_keys WHERE key_hash = ?",
        (candidate_hash,),
    ).fetchone()
    if row is None:
        return False
    return secrets.compare_digest(candidate_hash, row["key_hash"])


@app.route("/users/<user_id>/api-keys", methods=["POST"])
def create_api_key(user_id: str):
    plaintext_key = generate_api_key(user_id)
    return (
        jsonify(
            {
                "message": "Store this API key securely. It will not be shown again.",
                "api_key": plaintext_key,
            }
        ),
        201,
    )


if __name__ == "__main__":
    init_db()
    app.run(debug=False)